training/dataset.py

```diff
@@ -14,16 +16,82 @@ def __len__(self):
     def shuffle(self):
         np.random.shuffle(self.data)
 
-    def iterbatches(self, batch_size):
+    def iterbatches(self, batch_size: int):
         batch = 0
         for _ in range(self.num_batches(batch_size)):
             start = batch*batch_size
             end = min((batch+1)*batch_size, len(self.data))
             yield self.data[start:end]
             batch += 1
 
-    def split(self, frac):
+    def split(self, frac: float) -> Tuple["Dataset", "Dataset"]:
+        """Split dataset into two, using a fraction.
+
+
+        Parameters
+        ----------
+
+        frac: float
+            The proportion of the dataset to have to split on.
+
+        Returns
+        -------
+        tuple(Dataset, Dataset)
+            The split datasets, the first dataset containing the fractional percentage
+            and the second with the leftover.
+
+        """
+        if frac > 1.0 or frac < 0.0:
+            raise ValueError("frac must be between 0 and 1")
         split_idx = int(len(self.data)*frac)
         train_dataset = self.data[:split_idx]
         test_dataset = self.data[split_idx:]
         return Dataset(train_dataset), Dataset(test_dataset)
+
+    def random_split(self, frac: float) -> Tuple["Dataset", "Dataset"]:
+        """Split dataset into two, using a fraction. Shuffles data prior to splitting.
+
+
+        Parameters
+        ----------
+
+        frac: float
+            The proportion of the dataset to have to split on.
+
+        Returns
+        -------
+        tuple(Dataset, Dataset)
+            The split datasets, the first dataset containing the fractional percentage
+            and the second with the leftover.
+
+        """
+        self.shuffle()
+        return self.split(frac)
+
+    def indice_split(self, left: List[int], right: List[int]) -> Tuple["Dataset", "Dataset"]:
+        """Split dataset into two, providing the indices of .
+
+
+        Parameters
+        ----------
+
+        frac: float
+            The proportion of the dataset to have to split on.
+
+        Returns
+        -------
+        tuple(Dataset, Dataset)
+            The split datasets, the first dataset containing the fractional percentage
+            and the second with the leftover.
+
+        """
+        left = set(left)
+        right = set(right)
+        indices = set(range(len(self)))
+        if len(left.intersection(right)) > 0:
+            raise ValueError("Left and right indices contain overlap")
+        if len(left.union(right)) != len(indices) or len(left.union(right).difference(indices)) > 0:
+            raise ValueError("Indices provided don't match dataset indices")
+        left_split = [self.data[i] for i in left]
+        right_split = [self.data[i] for i in right]
+        return Dataset(left_split), Dataset(right_split)
```

Comment:
Nice!
* Could you mention in docstrings when to use `split` vs `random_split` vs `indice_split`?
* In addition to returning the result of the split, calling `dataset.random_split()` will modify `dataset` in place. Not sure if any later step might assume the order of items within `dataset` was the same as at initialization. If so, it might be safer instead to do something like `train_inds, validation_inds = ...; return Dataset(self.data[train_inds]), Dataset(self.data[validation_inds])`?